File: src/morie/fn/ksr13.py
# morie.fn — function file (hadesllm/morie)
"""Tangent-space dimension for parametric submodels (Kosorok 2008, Ch 6).

The tangent space T at P is the closure of {score functions s along
all smooth one-parameter paths through P}.  Its empirical analogue is
the rank of the centred-score covariance matrix.  We use the basis
of moment scores {x - mean(x), x^2 - mean(x^2)} and return its rank.
"""
import numpy as np
from ._richresult import RichResult

__all__ = ["kosorok_tangent_space"]


def kosorok_tangent_space(x):
    """Empirical tangent-space dimension via moment-score rank.

    Parameters
    ----------
    x : array-like.

    Returns
    -------
    RichResult with keys estimate (rank), n, method.
    """
    x = np.asarray(x, dtype=float)
    n = len(x)
    s1 = x - x.mean()
    s2 = x ** 2 - (x ** 2).mean()
    S = np.column_stack([s1, s2])
    G = (S.T @ S) / n
    rank = int(np.linalg.matrix_rank(G, tol=1e-10))
    return RichResult(payload={
        "estimate": rank,
        "n":        n,
        "method":   "Tangent-space dim via empirical-score rank",
    })


def cheatsheet():
    return "ksr13: tangent-space dimension via empirical-score rank"


# CANONICAL TEST
if __name__ == "__main__":
    rng = np.random.default_rng(0)
    print(kosorok_tangent_space(rng.normal(size=200)))

File: r-package/morie/R/ksr13.R
# SPDX-License-Identifier: GPL-2.0-only

#' Tangent-space dimension via empirical-score rank
#'
#' Score basis (x - mean(x)) and (x^2 - mean(x^2)); returns rank of
#' empirical Gram matrix at tolerance 1e-10.
#'
#' @param x Numeric vector.
#' @return Named list with estimate (rank), n, method.
#' @references Kosorok (2008), Ch 6.
#' @export
ksr13_kosorok_tangent_space <- function(x) {
  x <- as.numeric(x)
  n <- length(x)
  s1 <- x - mean(x)
  s2 <- x^2 - mean(x^2)
  S <- cbind(s1, s2)
  G <- crossprod(S) / n
  rank <- as.integer(qr(G, tol = 1e-10)$rank)
  list(
    estimate = rank,
    n        = n,
    method   = "Tangent-space dim via empirical-score rank"
  )
}

# CANONICAL TEST
# set.seed(0); ksr13_kosorok_tangent_space(rnorm(200))

#' @rdname ksr13_kosorok_tangent_space
#' @keywords internal
#' @export
kosorok_tangent_space <- ksr13_kosorok_tangent_space
